### PR TITLE
Closes #39: CoEQ test data

### LINKED ISSUE #39: Test data: CoEQ
### Background Information

We need to make test data for the control of eating questionnaire, so that we can use it to run templates/vignettes.

See discussion [here](https://github.com/pharmaverse/admiralmetabolic/issues/13).

On hold until we get confirmation of the license from @kathrinflunkert.

### Definition of Done

- [ ] Create `qs_metabolic` test dataset with Control of Eating questionnaire test answers. This should use same pts as in `dm_metabolic` and same visit schedule as in `vs_metabolic`.
- [ ] Documentation should include text to the effect of: _UoLeeds is the license holder and any parties wishing to use the CoEQ questionnaire need to purchase the license (at the appropriate link)_ (precise wording to be agreed with UoLeeds).

### PR DESCRIPTION
**Pull request by manciniedoardo**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral family codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral<ext> (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by manciniedoardo** [comment]
Hi @pharmaverse/admiralmetabolic - this is a PR to add a very simple CoEQ test dataset. I have added a couple of reviewers, but could anyone with capacity please take a look and see if the SDTM structure makes sense to you? I have not worked myself with the data and the only thing I had access to was some raw, unmapped data from the Roche side, so I had to "guess" the SDTM mapping! In particular, I also guessed the QSCAT/QSSCAT/QSTEST assignments, but of course we can modify them if needed.

@kathrinflunkert note that I have added the licensing disclaimer as well, which we'll need to send to UoLeeds before we merge in.

File: R/data.R [hunk 1]
```diff
@@ -11,3 +11,13 @@
 #'
 #' @source Constructed using `vs` from the `{pharmaversesdtm}` package and `dm_metabolic` from `{admiralmetabolic}` package
 "vs_metabolic"
+
+#' Example questionnaire dataset
+#'
+#' An example questionnaire SDTM dataset (containing Control of Eating Questionnaire, COEQ, test data) for metabolic studies. Note that University
+#' of Leeds are the copyright holders of the CoEQ and the test data included within `{admiralmetabolic}`is for not-for-profit use only. Any persons
```
Comment (AndersAskeland) [inline R/data.R:18]: Does the following imply that people can use the data we provide for non-profit activities? 
"... data included within `{admiralmetabolic}` is for not-for-profit use only."

And if that is the case, do we have confirmation from the copyright holder that this is OK?

File: data-raw/qs_metabolic.R [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#' Dataset: qs_metabolic
+#' Description: Create QS test SDTM dataset for metabolic studies,
+#' containing COEQ data (Control of Eating Questionnaire)
+#' Note that University  of Leeds are the copyright holders of the CoEQ and the test data included within
```
Comment (AndersAskeland) [inline data-raw/qs_metabolic.R:4]: See my comment in R/data.R

File: data-raw/qs_metabolic.R [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  "COEQ", "CONTROL EATING", "COEQ20", "Which one food makes it most difficult for you to control eating?",
+  "COEQ", "CONTROL EATING", "COEQ21", "How difficult has it been to resist eating this food during the last 7 days?",
+)
+
+# Use visit schedule and days from VS ----
+visit_schedule <- vs_metabolic %>%
+  select(STUDYID, USUBJID, VISIT, VISITNUM, VISITDY, VSDTC, VSDY) %>%
+  filter(!grepl("AMBUL|RETRIEVAL", VISIT)) %>%
+  rename(QSDTC = VSDTC, QSDY = VSDY) %>%
+  distinct()
```
Comment (AndersAskeland) [inline data-raw/qs_metabolic.R:54]: I would personally go with `stringr::str_detect()` vs `grepl()`. This change would also require us to load `stringr` (and remove inline namespace reference).

```suggestion
visit_schedule <- vs_metabolic %>%
  select(STUDYID, USUBJID, VISIT, VISITNUM, VISITDY, VSDTC, VSDY) %>%
  filter(stringr::str_detect(VISIT, "AMBUL|RETRIEVAL", negate = TRUE)) %>%
  rename(QSDTC = VSDTC, QSDY = VSDY) %>%
  distinct()
```

File: data-raw/qs_metabolic.R [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+# Cross join to get questions at each visit ----
+qs_metabolic_shell <- visit_schedule %>%
+  dplyr::cross_join(coeq_structure)
```
Comment (AndersAskeland) [inline data-raw/qs_metabolic.R:58]:
```suggestion
  cross_join(coeq_structure)
```
Unless there is a conflict with pharmaversesdtm or purrr?

File: data-raw/qs_metabolic.R [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+# Simulate question answers ----
+qs_metabolic_results <- qs_metabolic_shell %>%
+  mutate(
+    DOMAIN = "QS",
+    QSORRES = floor(runif(1) * 10) + round(runif(1)),
+    QSORRESU = "cm",
+    QSSTRESU = "cm",
+    QSSTRESN = QSORRES,
+    QSSTRESC = as.character(QSORRES),
+    QSBLFL = ifelse(VISIT == "BASELINE", "Y", NA_character_)
```
Comment (AndersAskeland) [inline data-raw/qs_metabolic.R:69]:
```suggestion
    QSBLFL = if_else(VISIT == "BASELINE", "Y", NA_character_)
```

**Review by AndersAskeland — CHANGES_REQUESTED** [review]
Thanks for the PR! I don't have much as I don't know the specific of actually using CoEQ.

I have a couple of things I would prefer being changed over to tidyverse, but besides that it looks good! :)

File: R/data.R (hunk 1 above)
Comment (manciniedoardo) [reply R/data.R:18]: hmm good point - no they would still need to request a not-for-profit use. What about:

Note that University of Leeds are the copyright holders of the CoEQ and the test data included within `{admiralmetabolic}` is for not-for-profit use only within `{admiralmetabolic}` and pharmaverse-related examples and documentation. Any persons or companies wanting to use the CoEQ should request a license to do so from the following [link](https://licensing.leeds.ac.uk/product/control-of-eating-questionnaire-coeq).

File: data-raw/qs_metabolic.R [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+#' is for not-for-profit use only within `{admiralmetabolic}` and `pharmaverse`-related examples/documentation. Any persons
+#' or companies wanting to use the CoEQ should request a license to do so from the following
+#' [link](https://licensing.leeds.ac.uk/product/control-of-eating-questionnaire-coeq).
+
+# Load libraries ----
+
+library(tibble)
+library(dplyr)
+library(pharmaversesdtm)
+library(purrr)
```
Comment (AndersAskeland) [inline data-raw/qs_metabolic.R:14]:
```suggestion
library(purrr)
library(stringr)
```

File: data-raw/qs_metabolic.R [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  "COEQ", "SPECIFIC CRAVINGS", "COEQ18", "Savoury foods (french fries, crisps, burgers, pizza, etc)",
+  "COEQ", "CONTROL EATING", "COEQ19", "Generally, how difficult has it been to control your eating?",
+  "COEQ", "CONTROL EATING", "COEQ20", "Which one food makes it most difficult for you to control eating?",
+  "COEQ", "CONTROL EATING", "COEQ21", "How difficult has it been to resist eating this food during the last 7 days?",
+)
+
+# Use visit schedule and days from VS ----
+visit_schedule <- vs_metabolic %>%
+  select(STUDYID, USUBJID, VISIT, VISITNUM, VISITDY, VSDTC, VSDY) %>%
+  filter(stringr::str_detect(VISIT, "AMBUL|RETRIEVAL", negate = TRUE)) %>%
```
Comment (AndersAskeland) [inline data-raw/qs_metabolic.R:52]:
```suggestion
  filter(str_detect(VISIT, "AMBUL|RETRIEVAL", negate = TRUE)) %>%
```

Moved to a library call to be consistent.

**Review by AndersAskeland — CHANGES_REQUESTED** [review]
I would propose this small change in library calls.

File: data-raw/qs_metabolic.R [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+# Set seed for random data generation ----
+set.seed(3.14159265)
+
+# Read input data ----
+
+vs_metabolic <- admiralmetabolic::vs_metabolic
+
+# Set up questions, codes and categories ----
+coeq_structure <- tibble::tribble(
+  ~QSCAT, ~QSSCAT, ~QSTESTCD, ~QSTEST,
```
Comment (kathrinflunkert) [inline data-raw/qs_metabolic.R:26]: I would omit QSSCAT. I don't think we would typically include such a grouping variable if it is not explicitly collected as such.

File: data-raw/qs_metabolic.R [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+# Cross join to get questions at each visit ----
+qs_metabolic_shell <- visit_schedule %>%
+  cross_join(coeq_structure)
+
+# Simulate question answers ----
+qs_metabolic_results <- qs_metabolic_shell %>%
+  mutate(
+    DOMAIN = "QS",
+    QSORRES = floor(runif(1) * 10) + round(runif(1)),
+    QSORRESU = "cm",
```
Comment (kathrinflunkert) [inline data-raw/qs_metabolic.R:66]: Unit should be mm.

File: data-raw/qs_metabolic.R [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+# Cross join to get questions at each visit ----
+qs_metabolic_shell <- visit_schedule %>%
+  cross_join(coeq_structure)
+
+# Simulate question answers ----
+qs_metabolic_results <- qs_metabolic_shell %>%
+  mutate(
+    DOMAIN = "QS",
+    QSORRES = floor(runif(1) * 10) + round(runif(1)),
```
Comment (kathrinflunkert) [inline data-raw/qs_metabolic.R:65]: Values should be from 0 to 100, apart from question 20 which is a free text question.

Comment (kathrinflunkert) [reply data-raw/qs_metabolic.R:65]: When I run this code it is not a random number that is generated but all rows get a value of QSORRES=2.

**Review by kathrinflunkert — CHANGES_REQUESTED** [review]

File: data-raw/qs_metabolic.R (hunk 10 above)
Comment (yurovska) [reply data-raw/qs_metabolic.R:65]:
```suggestion
    QSORRES = floor(runif(n()) * 10) + round(runif(n())),
```

File: data-raw/qs_metabolic.R [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+attr(qs_metabolic_seq$QSTESTCD, "label") <- "Questionnaire Test Short Name"
+attr(qs_metabolic_seq$QSORRES, "label") <- "Result or Finding in Original Units"
+attr(qs_metabolic_seq$QSORRESU, "label") <- "Original Units"
+attr(qs_metabolic_seq$QSSTRESC, "label") <- "Character Result/Finding in Std Format"
+attr(qs_metabolic_seq$QSSTRESN, "label") <- "Numeric Result/Finding in Standard Units"
+attr(qs_metabolic_seq$QSSTRESU, "label") <- "Standard Units"
+attr(qs_metabolic_seq$QSSEQ, "label") <- "Sequence Number"
+
+# Label QS dataset ----
+attr(qs_metabolic_seq, "label") <- "Questionnaire"
```
Comment (yurovska) [inline data-raw/qs_metabolic.R:99]:
```suggestion
attr(qs_metabolic_seq, "label") <- "Questionnaires"
```

File: data-raw/qs_metabolic.R [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+attr(qs_metabolic_seq$QSCAT, "label") <- "Category for Questionnaire"
+attr(qs_metabolic_seq$QSSCAT, "label") <- "Subcategory for Questionnaire"
+attr(qs_metabolic_seq$QSTEST, "label") <- "Questionnaire Test Name"
+attr(qs_metabolic_seq$QSTESTCD, "label") <- "Questionnaire Test Short Name"
+attr(qs_metabolic_seq$QSORRES, "label") <- "Result or Finding in Original Units"
+attr(qs_metabolic_seq$QSORRESU, "label") <- "Original Units"
+attr(qs_metabolic_seq$QSSTRESC, "label") <- "Character Result/Finding in Std Format"
+attr(qs_metabolic_seq$QSSTRESN, "label") <- "Numeric Result/Finding in Standard Units"
+attr(qs_metabolic_seq$QSSTRESU, "label") <- "Standard Units"
+attr(qs_metabolic_seq$QSSEQ, "label") <- "Sequence Number"
```
Comment (yurovska) [inline data-raw/qs_metabolic.R:96]:
```suggestion
labels <- list(
  DOMAIN = "Domain Abbreviation",
  QSBLFL = "Baseline Flag",
  QSCAT = "Category for Questionnaire",
  QSSCAT = "Subcategory for Questionnaire",
  QSTEST = "Questionnaire Test Name",
  QSTESTCD = "Questionnaire Test Short Name",
  QSORRES = "Result or Finding in Original Units",
  QSORRESU = "Original Units",
  QSSTRESC = "Character Result/Finding in Std Format",
  QSSTRESN = "Numeric Result/Finding in Standard Units",
  QSSTRESU = "Standard Units",
  QSSEQ = "Sequence Number"
)

for (var in names(labels)) {
  attr(qs_metabolic_seq[[var]], "label") <- labels[[var]]
}
```

File: data-raw/qs_metabolic.R [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+#' is for not-for-profit use only within `{admiralmetabolic}` and `pharmaverse`-related examples/documentation. Any persons
+#' or companies wanting to use the CoEQ should request a license to do so from the following
+#' [link](https://licensing.leeds.ac.uk/product/control-of-eating-questionnaire-coeq).
+
+# Load libraries ----
+
+library(tibble)
+library(dplyr)
+library(pharmaversesdtm)
+library(purrr)
```
[the remark this answers is not in the document]
Comment (manciniedoardo) [reply data-raw/qs_metabolic.R:14]: thanks, removed

File: data-raw/qs_metabolic.R (hunk 9 above)
Comment (manciniedoardo) [reply data-raw/qs_metabolic.R:66]: good spot, updated

File: data-raw/qs_metabolic.R (hunk 8 above)
Comment (manciniedoardo) [reply data-raw/qs_metabolic.R:26]: thanks, removed

**Comment by manciniedoardo** [comment]
Thanks @kathrinflunkert @yurovska for your reviews. I removed QSSCAT, ensured the answer to question 20 is categorical, and made the other suggested updates. Please review again - looking to merge by EoB Wed this week. Thanks!

File: data-raw/qs_metabolic.R [hunk 14]
```diff
@@ -62,19 +60,34 @@ qs_metabolic_shell <- visit_schedule %>%
...
+    QSSTRESU = "mm",
+    QSSTRESN = as.numeric(QSORRES),
+    QSSTRESC = QSORRES,
     QSBLFL = if_else(VISIT == "BASELINE", "Y", NA_character_)
+  ) %>%
+  # Update the free text question (COEQ20)
+  mutate(
+    QSSTRESC = case_when(
+      QSTESTCD == "COEQ20" & QSSTRESN < 33 ~ "Ice Cream",
+      QSTESTCD == "COEQ20" & QSSTRESN < 66 ~ "Pizza",
```
Comment (yurovska) [inline data-raw/qs_metabolic.R:74]: What if `QSSTRESN` for `COEQ18` [Savoury foods (french fries, crisps, burgers, pizza, etc)] turns out to be zero?

**Review by yurovska — APPROVED** [review]
@manciniedoardo Thanks for the updates. Just one pedantic comment, otherwise looks good to me.

File: data-raw/qs_metabolic.R (hunk 14 above)
Comment (manciniedoardo) [reply data-raw/qs_metabolic.R:74]: Haha fair point. Luckily, for this seed, it never is 😄 
image

**Review by AndersAskeland — APPROVED** [review]
LGTM! Thanks @manciniedoardo !

**Review by kathrinflunkert — APPROVED** [review]
Many thanks for the updates!

File: data-raw/qs_metabolic.R (hunk 10 above)
Comment (starosto) [reply data-raw/qs_metabolic.R]: @kathrinflunkert shouldn't we have a 0-10 scale for individual questions, and in the ADQS a 0-100 scale for subscales?

Comment (kathrinflunkert) [reply data-raw/qs_metabolic.R]: @starosto I am still trying to gather some more information about the different versions of the CoEQ, but the 21-item version for which we got the license has a 0 to 100 mm VAS scale:

  image



There is apparently another version where the user is only able to tick values between 0 and 10, but for this version I cannot find where/how to license it:
 
image

Comment (starosto) [reply data-raw/qs_metabolic.R]: @kathrinflunkert I see, I didn't know about the CoEQ VAS scale. Thanks for the clarification! :)
